### ISSUE TITLE
How to compile for Windows

### ISSUE THREAD

**Issue by C00LPANDA** [andreacasalino/Minimal-Socket#18]
I'm using Windows 11 for ARM VM with Parallels, on a M1 Pro MacBook Pro.

TL;DR: I just want to compile for Windows and I can't get my head around how am I supposed to do so.

I tried compiling with Cygwin, after I ran "make" I got:
```
In file included from /cygdrive/c/Users/username/Documents/Developer/C++/projectname/Minimal-Socket-master/src/src/Error.cpp:10:
/cygdrive/c/Users/username/Documents/Developer/C++/projectname/Minimal-Socket-master/src/src/SocketId.h:26:2: error: #error "Not supported system by MinimalSocket"
   26 | #error "Not supported system by MinimalSocket"
      |  ^~~~~
make[2]: *** [src/CMakeFiles/MinimalSocket.dir/build.make:76: src/CMakeFiles/MinimalSocket.dir/src/Error.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:177: src/CMakeFiles/MinimalSocket.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
Which makes sense because neither `_WIN32` nor `__linux__` are defined using Cygwin. 

I tried MSYS2, after compiling I got:
`'mutex' in namespace 'std' does not name a type`

I tried using MinGW binaries from this repository: https://github.com/niXman/mingw-builds-binaries/releases
When I use Windows threads and CMake (v3.24.1) I run: 
`C:\CMake\bin\cmake.exe ./CMakeLists.txt -G "MinGW Makefiles"`
`mingw32-make`

and I get the above mutex error.

When I use POSIX threads and CMake (v3.24.1) I run:
`C:\CMake\bin\cmake.exe ./CMakeLists.txt -G "MinGW Makefiles"`
`mingw32-make`

The library compiled successfully, I get a libMinimalSocket.a file instead of the expected .lib file, I think it's related to the POSIX thread model. Whenever I try to compile a program including the library, everything's fine but as soon as I try to use a class from it I get the following linker error:
`C:/Users/username/Desktop/mingw64POSIX/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\username\AppData\Local\Temp\cctkxgqD.o:C:\Users\username\Desktop\project/main.cpp:7: undefined reference to MinimalSocket::tcp::TcpServer::TcpServer(unsigned short, MinimalSocket::AddressFamily const&)'
C:/Users/username/Desktop/mingw64POSIX/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\username\AppData\Local\Temp\cctkxgqD.o: in function MinimalSocket::tcp::TcpServer::~TcpServer()':
C:\Users\username\Desktop\project/./Deps/MinimalSocket/tcp/TcpServer.h:37: undefined reference to MinimalSocket::Socket::~Socket()'
C:/Users/username/Desktop/mingw64POSIX/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\username\AppData\Local\Temp\cctkxgqD.o:main.cpp:(.rdata$.refptr._ZTTN13MinimalSocket3tcp9TcpServerE[.refptr._ZTTN13MinimalSocket3tcp9TcpServerE]+0x0): undefined reference to VTT for MinimalSocket::tcp::TcpServer'
C:/Users/username/Desktop/mingw64POSIX/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\username\AppData\Local\Temp\cctkxgqD.o:main.cpp:(.rdata$.refptr._ZTVN13MinimalSocket3tcp9TcpServerE[.refptr._ZTVN13MinimalSocket3tcp9TcpServerE]+0x0): undefined reference to vtable for MinimalSocket::tcp::TcpServer'
collect2.exe: error: ld returned 1 exit status
mingw32-make: *** [makefile:14: main] Error 1`

The file's content is just: `#include <MinimalSocket/tcp/TcpServer.h>
int main() {
    MinimalSocket::Port port = 12345;
    MinimalSocket::tcp::TcpServer tcpserver(port, MinimalSocket::AddressFamily::IP_V4);
}`

Compiled by: `g++ -std=c++20 -I./Deps -L./Deps/WindowsLib  -lMinimalSocket -g main.cpp -o main`

I became desperate and installed VS Build Tools for ARM, successfully compiled the library using their CMake by running `cmake --build .` it compiled to the expected .lib file.
Then I tried compiling server.cpp with MinGW, still the same linker error.
However, when I'm compiling with cl.exe I get a different linker error:
```
cl /std:c++20 /I.\Deps\ .\Deps\MinimalSocket.lib
server.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.34.31721\include\chrono(2206): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
Microsoft (R) Incremental Linker Version 14.34.31721.0
Copyright (C) Microsoft Corporation.  All rights reserved.
/out:server.exe
server.obj
.\Deps\MinimalSocket.lib
server.obj : error LNK2019: unresolved external symbol "public: virtual __thiscall MinimalSocket::Socket::~Socket(void)" (??1Socket@MinimalSocket@@UAE@XZ) referenced in function "public: void __thiscall MinimalSocket::tcp::TcpServer::vbase destructor'(void)" (??_DTcpServer@tcp@MinimalSocket@@QAEXXZ)
server.obj : error LNK2019: unresolved external symbol "public: __thiscall MinimalSocket::tcp::TcpServer::TcpServer(unsigned short,enum MinimalSocket::AddressFamily const &)" (??0TcpServer@tcp@MinimalSocket@@QAE@GABW4AddressFamily@2@@Z) referenced in function _main
.\Deps\MinimalSocket.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'x86'
server.exe : fatal error LNK1120: 2 unresolved externals
```
How else should I compile the project for Windows?

**Comment by andreacasalino** [comment]
The easiest way to compile in Windows is to use Visual Studio (community edition is completely free). CMake is able ti configure for you a Visual Studio project that you can use ti compile the targets you need.

Alternatively you can use the MinGW compiler.

Another way could be to download the artifacts (.lib or .DLL) from the GitHub actions page and then use the binaries in the way you prefer.

**Comment by andreacasalino** [comment]
Indeed from here 
https://github.com/andreacasalino/Minimal-Socket/actions/runs/2899140125 

windows-VS-... Are the folder storing the shared/static version of the library already compiled for you
